ibis/common/deferred.py

```diff
@@ -0,0 +1,556 @@
+from __future__ import annotations
+
+import collections.abc
+import functools
+import inspect
+import operator
+from abc import abstractmethod
+from typing import Any, Callable, TypeVar, overload
+
+from ibis.common.bases import FrozenSlotted as Slotted
+from ibis.common.bases import Hashable
+from ibis.common.collections import FrozenDict
+from ibis.util import is_iterable
+
+
+class Builder(Hashable):
```

Comment:
Hashable object describing the python operations used to construct the desired value. These don't support any sort of syntax sugar, they are rather similar to `ibis.expr.operations`.